corehq/apps/userreports/tasks.py

```diff
@@ -178,25 +186,52 @@ def _run_report(backend_to_use):
     return objects
 
 
+@periodic_task(
+    run_every=ASYNC_INDICATOR_QUEUE_TIME,
+    queue=settings.CELERY_PERIODIC_QUEUE,
+)
+def queue_async_indicators():
+    start = datetime.utcnow()
+    cutoff = start + ASYNC_INDICATOR_QUEUE_TIME
+    time_for_crit_section = ASYNC_INDICATOR_QUEUE_TIME.seconds - 10
+
+    with CriticalSection(['queue-async-indicators'], timeout=time_for_crit_section):
+        redis_client = get_redis_client().client.get_client()
+        indicators = AsyncIndicator.objects.all()[:10000]
+        paginator = Paginator(indicators, 1000)
+        for page in paginator.page_range:
+            for indicator in paginator.page(page).object_list:
+                now = datetime.utcnow()
+                if now > cutoff:
+                    break
+
+                lock_key = _get_indicator_queued_lock_key(indicator)
+                lock = redis_client.lock(lock_key, timeout=60 * 60 * 6)
+                if not lock.acquire(blocking=False):
+                    continue
+
+                save_document.delay(indicator.id, indicator.doc_id, indicator.pillow)
+
+
+def _get_indicator_queued_lock_key(indicator):
+    return 'async_indicator_queued-{}'.format(indicator.id)
+
+
 @task(queue=UCR_INDICATOR_CELERY_QUEUE, ignore_result=True, acks_late=True)
-def save_document(indicator_config_ids, doc, from_pillow_id):
-    error = PillowError.objects.get(doc_id=doc['_id'], pillow=from_pillow_id)
-    try:
-        for config_id in indicator_config_ids:
+def save_document(async_indicator_id, doc_id, pillow):
+    lock_key = get_async_indicator_modify_lock_key(doc_id, pillow)
+    with CriticalSection([lock_key]):
+        indicator = AsyncIndicator.objects.get(pk=async_indicator_id)
+        doc_store = get_document_store(indicator.domain, indicator.doc_type)
+        doc = doc_store.get_document(indicator.doc_id)
+
+        for config_id in indicator.indicator_config_ids:
             config = _get_config_by_id(config_id)
             adapter = get_indicator_adapter(config, can_handle_laboratory=True)
             adapter.best_effort_save(doc)
-    except Exception as exception:
-        error.add_attempt(exception, sys.exc_info()[2])
-        error.save()
-        error_id = error.id
-        pillow_logging.exception(
-            "[%s] Error on change: %s, %s. Logged as: %s" % (
-                from_pillow_id,
-                doc['_id'],
-                exception,
-                error_id
-            )
-        )
-    else:
-        error.delete()
+
+        redis_client = get_redis_client().client.get_client()
+        queued_lock_key = _get_indicator_queued_lock_key(indicator)
+        lock = redis_client.lock(queued_lock_key)
+        release_lock(lock, degrade_gracefully=True)
```

Comment:
The lock has a timeout https://github.com/dimagi/commcare-hq/pull/15462/files#diff-1f15a4ea1fec5c4ee5fa155d4d80c3d7R208 so that will time out and it will be requeued in celery